java/ql/src/semmle/code/java/security/ExternalAPIs.qll

```diff
@@ -0,0 +1,146 @@
+      getName() = "isNullOrEmpty"
+      or
+      getDeclaringType().hasQualifiedName("org.apache.commons.lang3", "StringUtils") and
+      getName() = "isNotEmpty"
+      or
+      getDeclaringType().hasQualifiedName("java.lang", "Character") and
+      getName() = "isDigit"
+      or
+      getDeclaringType().hasQualifiedName("java.lang", "String") and
+      getName().regexpMatch("equalsIgnoreCase|regionMatches")
+      or
+      getDeclaringType().hasQualifiedName("java.lang", "Boolean") and
+      getName() = "parseBoolean"
+      or
+      getDeclaringType().hasQualifiedName("org.apache.commons.io", "IOUtils") and
+      getName() = "closeQuietly"
+      or
+      getDeclaringType().hasQualifiedName("org.springframework.util", "StringUtils") and
+      getName().regexpMatch("hasText|isEmpty")
+    }
+  }
+
+  /** A node representing data being passed to an external API. */
+  class ExternalAPIDataNode extends DataFlow::Node {
+    Call call;
+    int i;
+
+    ExternalAPIDataNode() {
+      (
+        // Argument to call to a method
+        this.asExpr() = call.getArgument(i)
+        or
+        // Qualifier to call to a method which returns non trivial value
+        this.asExpr() = call.getQualifier() and
+        i = -1 and
+        not call.getCallee().getReturnType() instanceof VoidType and
+        not call.getCallee().getReturnType() instanceof BooleanType
+      ) and
+      // Defined outside the source archive
+      not call.getCallee().fromSource() and
+      // Not a call to an method which is overridden in source
+      not exists(Method m |
+        m.getASourceOverriddenMethod() = call.getCallee().getSourceDeclaration() and
+        m.fromSource()
+      ) and
+      // Not already modeled as a taint step
+      not exists(DataFlow::Node next | TaintTracking::localTaintStep(this, next)) and
+      // Not a call to a known safe external API
+      not call.getCallee() instanceof SafeExternalAPIMethod
+    }
+
+    /** Gets the called API `Method`. */
+    Method getMethod() { result = call.getCallee() }
+
+    /** Gets the index which is passed untrusted data (where -1 indicates the qualifier). */
+    int getIndex() { result = i }
+
+    /** Gets the description of the method being called. */
+    string getMethodDescription() {
+        result =
+          getMethod().getDeclaringType().getPackage() + "." + getMethod().getQualifiedName()
+    }
+  }
+
+  class UntrustedDataToExternalAPIConfig extends TaintTracking::Configuration {
+    UntrustedDataToExternalAPIConfig() { this = "UntrustedDataToExternalAPIConfig" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof ExternalAPIDataNode }
+  }
+
+  /** A node representing untrusted data being passed to an external API. */
+  class UntrustedExternalAPIDataNode extends ExternalAPIDataNode {
```

Comment:
```suggestion
  private class UntrustedExternalAPIDataNode extends ExternalAPIDataNode {
```

Reply:
The `ExternalAPIUsedWithUntrustedData` class, which is used in the `ExternalAPIsUsedWithUntrustedData.ql` query, exposes this through the `getUntrustedDataNode()` member predicate, so I think it makes sense to keep it public as well.

Reply:
Right, I missed that.